portalnet/src/offer_queue.rs

```diff
@@ -0,0 +1,313 @@
+use crate::overlay::{
+    command::OverlayCommand,
+    request::{OverlayRequest, RequestDirection},
+};
+use delay_map::HashMapDelay;
+use ethportal_api::{
+    types::{
+        enr::Enr,
+        portal_wire::{FindContent, Request},
+    },
+    OverlayContentKey,
+};
+use futures::prelude::*;
+use rand::{seq::SliceRandom, thread_rng};
+use tokio::{sync::mpsc::UnboundedSender, time::Duration};
+use tracing::{debug, warn};
+
+struct SeenPeers {
+    origin: Enr,
+    fallback: Vec<Enr>,
+}
+
+/// In-memory queue of content keys that have been accepted, and are
+/// currently being processed.
+/// This queue is designed for content keys that have already
+/// been checked with storage that they are within radius.
+// It's possible that some content keys evade tracking through the entire processing
+// cycle, due to error handling inside the OverlayService (eg process_accept_utp_payload()).
+// On top of this, we could also have threads panic / deadlocking. The use of a delay
+// map is useful to remove content keys which have been in the queue for too long or
+// might be stuck, and avoid infinitely-growing queues / indefinitely prevent
+// previously seen content keys from being accepted.
+pub struct OfferQueue<TContentKey>
+where
+    TContentKey: OverlayContentKey,
+{
+    // a map of content keys actively being transferred / processed
+    // pointing to seen peers that have offered them the content
+    content_key_map: HashMapDelay<TContentKey, SeenPeers>,
+    command_tx: UnboundedSender<OverlayCommand<TContentKey>>,
+}
+
+const OFFER_QUEUE_TIMEOUT: Duration = Duration::from_secs(120);
+
+impl<TContentKey> OfferQueue<TContentKey>
+where
+    TContentKey: OverlayContentKey,
+{
+    pub fn new(command_tx: UnboundedSender<OverlayCommand<TContentKey>>) -> Self {
+        Self {
+            content_key_map: HashMapDelay::new(OFFER_QUEUE_TIMEOUT),
+            command_tx,
+        }
+    }
+
+    // Checks if we should accept the given content key based on the queue state.
+    // - Polls for expired items, which will remove them from the queue.
+    // - If the key is not in the queue, it is added to the queue and returns true.
+    // - If the key is in the queue, the seen peer is stored, and it returns false.
+    pub fn should_accept(&mut self, content_key: &TContentKey, peer: &Enr) -> bool {
+        // poll for expired items, which will remove them from the queue
+        let _ = future::poll_fn(|cx| self.content_key_map.poll_expired(cx)).now_or_never();
+        if let Some(mut node_ids) = self.content_key_map.remove(content_key) {
+            if node_ids.origin == *peer || node_ids.fallback.contains(peer) {
+                warn!(
+                    "Received multiple offers containing the same content key: {} from peer: {}",
+                    content_key,
+                    peer.node_id(),
+                );
+                return false;
+            }
```

Comment:
```
self.content_key_map.remove(content_key) {
```
We are removing the key shouldn't we have `self.content_key_map.insert(content_key.clone(), node_ids);` before the return

Reply:
Ahh yeah, good catch. I also added a test case to cover this